libraries/plugins/token_emissions/token_emissions_plugin.cpp

```diff
@@ -0,0 +1,173 @@
+#include <steem/chain/index.hpp>
+#include <steem/chain/smt_objects.hpp>
+#include <steem/chain/util/smt_token.hpp>
+#include <steem/protocol/config.hpp>
+
+#include <fc/io/json.hpp>
+#include <cmath>
+
+namespace steem { namespace plugins { namespace token_emissions {
+
+namespace detail {
+
+class token_emissions_impl
+{
+public:
+   token_emissions_impl() : _db( appbase::app().get_plugin< steem::plugins::chain::chain_plugin >().db() ) {}
+   virtual ~token_emissions_impl() {}
+
+   void on_post_apply_required_action( const required_action_notification& note );
+   void on_post_apply_optional_action( const optional_action_notification& note );
+   void on_generate_optional_actions( const generate_optional_actions_notification& note );
+
+   chain::database&              _db;
+   boost::signals2::connection   post_apply_required_action_connection;
+   boost::signals2::connection   post_apply_optional_action_connection;
+   boost::signals2::connection   generate_optional_action_connection;
+};
+
+void token_emissions_impl::on_post_apply_required_action( const required_action_notification& note )
+{
+   if ( note.action.which() != required_automated_action::tag< smt_token_launch_action >::value )
+      return;
+
+   if ( _db.is_pending_tx() )
+      return;
+
+   smt_token_launch_action token_launch = note.action.get< smt_token_launch_action >();
+
+   auto next_emission = util::smt::next_emission_time( _db, token_launch.symbol );
+
+   if ( next_emission )
+   {
+      _db.create< token_emission_schedule_object >( [ token_launch, next_emission ]( token_emission_schedule_object& o )
+      {
+         o.symbol = token_launch.symbol;
+         o.next_consensus_emission = *next_emission;
+         o.next_scheduled_emission = *next_emission;
+      } );
+   }
+}
+
+void token_emissions_impl::on_post_apply_optional_action( const optional_action_notification& note )
+{
+   if ( note.action.which() != optional_automated_action::tag< smt_token_emission_action >::value )
+      return;
+
+   if ( _db.is_pending_tx() )
+      return;
+
+   smt_token_emission_action emission_action = note.action.get< smt_token_emission_action >();
+
+   auto next = util::smt::next_emission_time( _db, emission_action.symbol, _db.get< smt_token_object, steem::chain::by_symbol >( emission_action.symbol ).last_virtual_emission_time );
+
+   const auto& emission_obj = _db.get< token_emission_schedule_object, by_symbol >( emission_action.symbol );
+   if ( next )
+   {
+      _db.modify( emission_obj, [ next ]( token_emission_schedule_object& o )
+      {
+         o.next_consensus_emission = *next;
+         o.next_scheduled_emission = *next;
+      } );
+   }
+   else
+   {
+      _db.remove( emission_obj );
+   }
+}
+
+void token_emissions_impl::on_generate_optional_actions( const generate_optional_actions_notification& note )
+{
+   // We're creating token emissions for the upcoming block.
+   time_point_sec next_emission_time = _db.head_block_time() + STEEM_BLOCK_INTERVAL;
```

Comment:
If we are generating on the opt action signal, we shouldn't need to add a block interval.